File: lv1-library-loan/csharp/LibraryLoan.Domain/Members/LoanRecord.cs
using LibraryLoan.BuildingBlocks;
using LibraryLoan.Domain.Books;
using LibraryLoan.Results;

namespace LibraryLoan.Domain.Members;

public sealed class LoanRecord : Entity<LoanRecordId>
{
    private const int LoanPeriodDays = 14;

    public BookId BookId { get; }
    public DateOnly LoanDate { get; }
    public DateOnly DueDate { get; }
    public LoanStatus Status { get; private set; }

    private LoanRecord(LoanRecordId id, BookId bookId, DateOnly loanDate, DateOnly dueDate, LoanStatus status)
        : base(id)
    {
        BookId = bookId;
        LoanDate = loanDate;
        DueDate = dueDate;
        Status = status;
    }

    public static LoanRecord Loan(LoanRecordId id, BookId bookId, DateOnly loanDate)
    {
        return new LoanRecord(id, bookId, loanDate, loanDate.AddDays(LoanPeriodDays), LoanStatus.Borrowed);
    }

    public bool IsOverdue(DateOnly asOfDate) => Status == LoanStatus.Borrowed && asOfDate > DueDate;

    public Result Return()
    {
        if (Status != LoanStatus.Borrowed)
        {
            return Result.Failure(LoanRecordErrors.AlreadyReturned(Id));
        }

        Status = LoanStatus.Returned;
        return Result.Success();
    }
}

File: lv1-library-loan/typescript/src/domain/members/loan-record.ts
import { Entity } from "../../building-blocks/entity.js";
import { Result } from "../../results/result.js";
import type { BookId } from "../books/book-id.js";
import type { LocalDate } from "../shared/local-date.js";
import { LoanRecordErrors } from "./loan-record-errors.js";
import type { LoanRecordId } from "./loan-record-id.js";
import { LoanStatus } from "./loan-status.js";

const LOAN_PERIOD_DAYS = 14;

export class LoanRecord extends Entity<LoanRecordId> {
  readonly bookId: BookId;
  readonly loanDate: LocalDate;
  readonly dueDate: LocalDate;
  private loanStatus: LoanStatus;

  private constructor(id: LoanRecordId, bookId: BookId, loanDate: LocalDate, dueDate: LocalDate, status: LoanStatus) {
    super(id);
    this.bookId = bookId;
    this.loanDate = loanDate;
    this.dueDate = dueDate;
    this.loanStatus = status;
  }

  get status(): LoanStatus {
    return this.loanStatus;
  }

  static loan(id: LoanRecordId, bookId: BookId, loanDate: LocalDate): LoanRecord {
    return new LoanRecord(id, bookId, loanDate, loanDate.addDays(LOAN_PERIOD_DAYS), LoanStatus.Borrowed);
  }

  isOverdue(asOfDate: LocalDate): boolean {
    return this.loanStatus === LoanStatus.Borrowed && asOfDate.isAfter(this.dueDate);
  }

  return(): Result<void> {
    if (this.loanStatus !== LoanStatus.Borrowed) {
      return Result.failure(LoanRecordErrors.alreadyReturned(this.id));
    }

    this.loanStatus = LoanStatus.Returned;
    return Result.success();
  }
}
